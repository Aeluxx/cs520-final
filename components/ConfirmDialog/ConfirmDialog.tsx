import React, { ReactNode } from 'react'
import { Dialog, Box, Typography, Button, Grid } from '@mui/material'

export type ConfirmDialogProps = {
  title?: ReactNode
  content?: ReactNode
  cancelText?: ReactNode
  confirmText?: ReactNode
  open?: boolean
  onClose?: VoidFunction
  onConfirm?: VoidFunction
  onCancel?: VoidFunction
  stackButtons?: boolean
}

const ConfirmDialog: React.FC<ConfirmDialogProps> = ({
  title = 'Confirmation Needed',
  content = (
    <>
      <strong>Are you sure?</strong>
    </>
  ),
  cancelText = <>Cancel</>,
  confirmText = <>Confirm</>,
  open = false,
  stackButtons = false,
  onClose = () => {},
  onConfirm = () => {},
  onCancel = () => {},
}) => {
  return (
    <Dialog open={open} onClose={onClose} maxWidth='xs'>
      <Box p={3}>
        <Box>
          {typeof title === 'string' ? (
            <Typography color='primary'>
              <strong>{title}</strong>
            </Typography>
          ) : (
            title
          )}
        </Box>
        <Box mt={2} mb={3}>
          {typeof content === 'string' ? <Typography>{content}</Typography> : content}
        </Box>
        <Box>
          <Grid container spacing={2} alignItems='stretch'>
            {stackButtons ? null : (
              <Grid item xs={5}>
                <Button fullWidth onClick={onCancel}>
                  {cancelText}
                </Button>
              </Grid>
            )}
            <Grid item xs={stackButtons ? 12 : 7}>
              <Button fullWidth color='primary' variant='contained' onClick={onConfirm}>
                {confirmText}
              </Button>
            </Grid>
            {stackButtons && (
              <Grid item xs={12}>
                <Button fullWidth onClick={onCancel}>
                  {cancelText}
                </Button>
              </Grid>
            )}
          </Grid>
        </Box>
      </Box>
    </Dialog>
  )
}

export default ConfirmDialog
