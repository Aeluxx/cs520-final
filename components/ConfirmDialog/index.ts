export { default } from './ConfirmDialog'
export { default as useConfirmDialog } from './useConfirmDialog'