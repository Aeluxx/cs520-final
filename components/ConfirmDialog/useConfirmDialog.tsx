import React, { useState, useCallback, useMemo, useRef } from "react";

import ConfirmDialog, { ConfirmDialogProps } from "./ConfirmDialog";

type Options = {
  defaultOpen?: boolean;
};

const useConfirmDialog = (options: Options = {}) => {
  const { defaultOpen = false } = options;
  const [open, setOpen] = useState<boolean>(defaultOpen);
  const show = useCallback(() => setOpen(true), [setOpen]);
  const hide = useCallback(() => setOpen(false), [setOpen]);
  const onConfirm = useRef<VoidFunction>(hide);
  const onCancel = useRef<VoidFunction>(hide);

  const confirm = useCallback(
    (onCancelArg: VoidFunction = () => {}) =>
      new Promise<boolean>((resolve) => {
        show();
        onConfirm.current = () => {
          resolve(true);
          hide();
          onConfirm.current = hide;
        };
        onCancel.current = () => {
          resolve(false);
          hide();
          onCancelArg();
          onCancel.current = hide;
        };
      }),
    [show, hide]
  );

  const render = useCallback(
    (props: ConfirmDialogProps = {}) => (
      <ConfirmDialog
        {...props}
        open={open}
        onClose={hide}
        onConfirm={onConfirm.current}
        onCancel={onCancel.current}
      />
    ),
    [open, hide]
  );

  return useMemo(
    () => ({
      render,
      confirm
    }),
    [render, confirm]
  );
};

export default useConfirmDialog;
