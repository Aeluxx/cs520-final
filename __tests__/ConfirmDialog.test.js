import React from 'react';
import '@testing-library/jest-dom';
import ConfirmDialog from '../components/ConfirmDialog/ConfirmDialog';
import ShallowRenderer from 'react-test-renderer/shallow';

let realUseContext;
let useContextMock;

beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
});

afterEach(() => {
    React.useContext = realUseContext;
});

it("ConfirmDialog renders without error", () => {
    useContextMock.mockReturnValue("Test Value");
    const shallow = new ShallowRenderer();
    const confirmDialog = shallow.render(<ConfirmDialog />);
});

it("ConfirmDialog matches Snapshot", () => {
    useContextMock.mockReturnValue("Test Value");
    const shallow = new ShallowRenderer();
    const confirmDialog = shallow.render(<ConfirmDialog />);
    const result = shallow.getRenderOutput();
    expect(result).toMatchSnapshot();
});