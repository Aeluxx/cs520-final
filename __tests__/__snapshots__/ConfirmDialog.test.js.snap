// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog matches Snapshot 1`] = `
<ForwardRef(Dialog)
  maxWidth="xs"
  onClose={[Function]}
  open={false}
>
  <ForwardRef(Box)
    p={3}
  >
    <ForwardRef(Box)>
      <ForwardRef(Typography)
        color="primary"
      >
        <strong>
          Confirmation Needed
        </strong>
      </ForwardRef(Typography)>
    </ForwardRef(Box)>
    <ForwardRef(Box)
      mb={3}
      mt={2}
    >
      <React.Fragment>
        <strong>
          Are you sure?
        </strong>
      </React.Fragment>
    </ForwardRef(Box)>
    <ForwardRef(Box)>
      <ForwardRef(Grid)
        alignItems="stretch"
        container={true}
        spacing={2}
      >
        <ForwardRef(Grid)
          item={true}
          xs={5}
        >
          <ForwardRef(Button)
            fullWidth={true}
            onClick={[Function]}
          >
            <React.Fragment>
              Cancel
            </React.Fragment>
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
        <ForwardRef(Grid)
          item={true}
          xs={7}
        >
          <ForwardRef(Button)
            color="primary"
            fullWidth={true}
            onClick={[Function]}
            variant="contained"
          >
            <React.Fragment>
              Confirm
            </React.Fragment>
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
      </ForwardRef(Grid)>
    </ForwardRef(Box)>
  </ForwardRef(Box)>
</ForwardRef(Dialog)>
`;
